nemo/collections/asr/models/rnnt_models.py

```diff
@@ -242,6 +242,10 @@ def transcribe(
         """
         if paths2audio_files is None or len(paths2audio_files) == 0:
             return {}
+        if return_hypotheses and self.cfg.decoding.return_encoder_embeddings:
```

Comment:
Wait, why is this being done as such ? 
The hypothesis object can hold any tensor. Make a new value like "encoder_embedding" : Optional[torch Tensor] = None. 

In greedy and beam decoding scripts, you use this new flag - rename to "preserve_encoder_embedding" - to check if encoder embeddings should be preserved or not. 

If yes, inside the forward() pass of each of the beam, greedy and batch greedy decoding classes, you slice off the appropriate encoder embedding for each sample, call detach and CPU() and then set them into the hypothesis object. 

Then user just sets this flag and does return_hypotheses = True - and the hypothesis object will contain the actual encoder embedding for that sample.

Reply:
Hi!
I am trying to implement the same functionality as in `ctc_models.trascribe()` -- it returns a list of decoded transcriptions or only raw logprobs from `ctc_encoder` (without decoded transcriptions -- https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/ctc_models.py#L202).

Raw logprobs are used for search of decoder hyper parameters by calling `model.decoding.ctc_decoder_predictions_tensor` method in the search loop. 

Do I understand correctly what your suggestion with `preserve_encoder_embedding` will do all the rnnt decoding pipeline (including getting decoded transcripts) and save encoder_embeddings additionally? I wanted to avoid transcripts decoding and return encoder_embeddings only (to save time) for later hyper parameters search by `model.decoding.rnnt_decoder_predictions_tensor`.

Reply:
The logprobs flag in CTC models transcribe is legacy from before we had Hypothesis object. These days you are meant to request return_hypotheses true in order to get the hypothesis object with the "alignment" field containing the logprobs. We won't be adding an equivalent field in RNNT because we want to follow the proper way of using decoding strategy and hypothesis object. 

In RNNT, alignment is decoder output - which we don't need. So for this script you can manually calculate the encoder forward pass and skip the decoder and joint for caching step, then use decoding . Rnnt_docoder_prediction_tensod from the cached inputs in two steps. No need to add flags then

Reply:
Ok, I will try to avoid the call of `rnnt_model.transcribe()` for encoder_embeddings calculation. I will do it manually inside the `eval_beamsearch_ngram_transducer.py` script.

Reply:
Thanks, this is good enough for now. Let's discuss offline for a better design so that we can reuse transcribe if possible